### DISCUSSION TITLE
密码功能希望增加"忽略"上层密码的选项

### DISCUSSION THREAD

**Discussion by jsjcjsjc** [spencerwooo/onedrive-vercel-index#844] [Q&A]
你好
感谢你的应用,非常棒!
有个需求希望可以考虑,就是目前的设置一个目录的密码后,访问下层所有文件都需要输入密码.
是否可以增加一个选项,只在界面上进入目录时需要输入密码,直连的方式可以忽略密码?

感谢

**Comment by spencerwooo** [comment]
（我把你发重复的那个删掉了。）

> 是否可以增加一个选项，只在界面上进入目录时需要输入密码，直连的方式可以忽略密码？

请看：[受密码保护的文件夹 - 注意事项](https://ovi.swo.moe/zh/docs/features/protected-folders#%E6%B3%A8%E6%84%8F%E4%BA%8B%E9%A1%B9)，在密码保护的文件夹里复制的直链是带有参数 `&odpt=<hashed_token>` 的，实际上就是把你的设置的密码经过混淆后放在链接里了，所以应该已经是有你说的这个功能了。

**Reply by jsjcjsjc** [reply]
> 

感谢,这个功能我也测试了,应该说是实现同一目的的两种方式.
我的想法可能稍微好看一点,直连后面不需要加`&odpt=c8ebdc273d673d167c238dd25a997bdf9b4ecbf589452697c1a6cb416b2ed11d`这样的后缀.
**before**
`https://drive.swo.moe/api/raw/?path=/Shared%20files/genshin-impact-npc-scene-interaction.mp4&odpt=c8ebdc273d673d167c238dd25a997bdf9b4ecbf589452697c1a6cb416b2ed11d`

**after**
`https://drive.swo.moe/api/raw/?path=/Shared%20files/genshin-impact-npc-scene-interaction.mp4`

而且暴露了`&odpt=c8ebdc273d673d167c238dd25a997bdf9b4ecbf589452697c1a6cb416b2ed11d`,可能反而不安全.
考虑一下哈,同学.
已经送上月卡 ~*(￣▽￣)／

**Reply by jsjcjsjc** [reply]
其实我是想做图床的,但是又不想让别人知道所有的图片,所以想在门口加个口令~~
如果不准备干,也非常感谢你的应用.

**Comment by spencerwooo** [comment]
谢谢赞助 @jsjcjsjc 🤗

我简单说一下为什么这里是这样的吧，因为密码保护是去年年初实现的功能了。

首先，密码保护是这样实现的：

- **项目后端** 从 OneDrive 那里请求 API 拿到当前目录下的所有文件，并试图直接请求目录下名为 `.password` 的文件；
- 如果发现其中有名为 `.password` 的文件，视为当前路径有密码，密码内容在 `.password` 文件里声明；
- 此时 **项目后端** 会直接返回一个「需要密码才可以」的 response，相当于告诉 **项目前端** 当前路径有密码；
- 这样，前端用户输入密码之后，重新发起请求，后端拿到密码，跟 `.password` 里存储的密码做校验，成功才会返回受保护的文件列表；

这里面在「前端用户输入密码之后，重新发起请求」中，「密码」是放在请求的 header 里的，这也就要求在前端请求所有受密码保护的文件夹里的内容时，都需要带有这个 header，或者用其他方法（比如带上这个 `odpt` 的 token），将「密码」告诉后端的 —— 否则加密码就失去意义了。这里包括直接列出文件列表，也包括请求受保护的文件内容。

而之所以考虑「直链」用 URL 参数 `odpt`，就恰好是考虑到在比如图床的场景下，用户是没有办法直接给 URL 加上请求 header 的，因为只是简单的插入一个直链链接。（在之前的实现里面，受密码保护的文件是无法生成直链的。）

**Comment by spencerwooo** [comment]
总之，现在在 URL 上加 `odpt` 其实是去年的一个折中的设计了，我个人估计是不会对这部分的逻辑做改动了。如果图床的话我其实建议直接用公开文件夹吧，因为本来就是要公开的内容，加一道密码并没有发挥太大作用呢。

**Reply by jsjcjsjc** [reply]
感谢回复
